compiler/rustc_middle/src/mir/interpret/allocation.rs

```diff
@@ -273,6 +286,20 @@ impl<Prov: Provenance> Allocation<Prov> {
             extra: (),
         })
     }
+
+    pub fn scalar_pair<'tcx>(
+        a: Scalar<Prov>,
+        b: Scalar<Prov>,
+        size: Size,
+        align: Align,
+        tcx: TyCtxt<'tcx>,
+    ) -> InterpResult<'tcx, Self> {
+        let mut ptr = Allocation::uninit(size, align, true)?;
+        ptr.write_scalar(&tcx, AllocRange { start: Size::ZERO, size: a.size() }, a).unwrap();
+        ptr.write_scalar(&tcx, AllocRange { start: a.size(), size: b.size() }, b).unwrap();
+        ptr.mutability = Mutability::Not;
+        Ok(ptr)
```

Comment:
I don't think we want to duplicate the subtle ScalarPair writing logic here. The above is actually wrong, the 2nd field needs to be at `a.size().align_to(b.align())`.

`alloc_type_name` should probably go through `op_to_const` (I guess the `ScalarPair` arm is reachable then.)
And `op_to_const` should use `write_immediate`.

Reply:
Okay yeah after reading a bit more of this... I think
- we want `op_to_const` to indeed handle all kinds of operands, so when it gets an immediate ScalarPair it needs to create a new allocation
- it should do that using the existing ScalarPair writing codepath, via `write_immediate`
- `alloc_type_name` should go through `op_to_const`

Reply:
... or even better can MIR building and `alloc_type_name` use valtrees? :D I guess that would need optimized valtree representations for `str` and `[u8]`. But then the `Immediate::ScalarPair` case in `op_to_const` should be unreachable, right?

Reply:
> even better can MIR building and `alloc_type_name` use valtrees?

yes, but not in this PR XD

Reply:
Yeah... I wondered about the best order to do this. If it's just about fixing that bug, we can do that easier -- we can keep `ConstValue::Slice` but just remove str and slices from the `try_as_immediate` computation.

But I guess if the complexity of creating that allocation for ScalarPair is contained in `op_to_const` then this is still a win overall.

Reply:
See 355f165ecee1eb10d0475659e3989b449d10340c for a version of this. Not ideal, but it doesn't break through any abstractions ^^